lib/intersection-type.helper.ts

```diff
@@ -6,49 +6,38 @@ import {
   inheritValidationMetadata,
 } from './type-helpers.utils';
 
-export function IntersectionType<A, B>(
-  target: Type<A>,
-  source: Type<B>,
-): MappedType<A & B>;
+// https://stackoverflow.com/questions/50374908/transform-union-type-to-intersection-type
+type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (
+  k: infer I,
+) => void
+  ? I
+  : never;
 
-export function IntersectionType<A, B, C>(
-  target: Type<A>,
-  sourceB: Type<B>,
-  sourceC: Type<C>,
-): MappedType<A & B & C>;
+type ClassRefsToConstructors<T> = {
+  [U in keyof T]: T[U] extends Type<infer V> ? V : never;
+};
 
-export function IntersectionType<A, B, C, D>(
-  target: Type<A>,
-  sourceB: Type<B>,
-  sourceC: Type<C>,
-  sourceD: Type<D>,
-): MappedType<A & B & C & D>;
-
-export function IntersectionType<A, T extends { new (...arg: any): any }[]>(
-  classA: Type<A>,
-  ...classRefs: T
-): MappedType<A> {
-  const allClassRefs = [classA, ...classRefs];
+type Intersection<T extends Type[]> = Type<
+  UnionToIntersection<ClassRefsToConstructors<T>[number]>
+>;
```

Comment:
This looks like a beastly type! Can I ask what happens when we have classes that have the same key? How do the types handle that?
e.g.
```ts
class Foo {
  key: string;
  foo: string;
}

class Bar {
  key: number;
  bar: string;
}

class FooBar extends IntersectionType(Foo, bar) {}
```

I'm worried that in a case like this, if one were to ever arise, we'd get an issue of `FooBar` not ever properly extending both `Foo` and `Bar` because of the implicit intersection of `string & number` which is `never`.

---

Other than that concern, the type looks great. Some comments to help understand it better could be nice, but not immediately necessary

Reply:
> ... what happens when we have classes that have the same key? How do the types handle that?
> ... the implicit intersection of string & number which is never

Yes, the type of the key `key` will be `never`, and it is the original behavior in the current version:
![螢幕截圖 2022-09-30 上午10 41 57](https://user-images.githubusercontent.com/57888887/193178946-f574ae41-9932-461b-ac52-4e889078104f.png)

Since the current returned type comes from 
https://github.com/nestjs/mapped-types/blob/master/lib/intersection-type.helper.ts#L12

Which returns `MappedType< Foo & Bar >`, and the key of `key` in `Foo & Bar` is `never`.

Reply:
> Some comments to help understand it better could be nice
Added description of how it works in comments :)